server/src/ActionManager.cc

```diff
@@ -549,6 +549,50 @@ void ActionManager::checkEvents(const EventInfoList &eventList)
 	}
 }
 
+void ActionManager::reExecuteUnfinishedAction(void)
+{
+	ThreadLocalDBCache cache;
+	DBTablesAction &dbAction = cache.getAction();
+	DBTablesMonitoring &dbMonitoring = cache.getMonitoring();
+
+	ActionLogList actionLogList;
+	vector<ActionLogStatus> targetStatuses{ACTLOG_STAT_QUEUING,
+	                                       ACTLOG_STAT_STARTED,
+	                                       ACTLOG_STAT_RESIDENT_QUEUING,
+	                                       ACTLOG_STAT_LAUNCHING_RESIDENT};
+
+	if (!dbAction.getTargetStatusesLogs(actionLogList, targetStatuses)) {
+		MLPL_INFO("Hatohol does not have unfinished action.\n");
+		return;
+	}
+
+	ActionLogListIterator actLogIt = actionLogList.begin();
+	for (; actLogIt != actionLogList.end(); ++actLogIt) {
+		ActionLog &actionLog = *actLogIt;
+		EventInfoList eventList;
+		EventsQueryOption eventOption(USER_ID_SYSTEM);
+
+		eventOption.setTargetServerId(actionLog.serverId);
+		string eventId = actionLog.eventId;
+		list<EventIdType> eventIds{eventId};
+		eventOption.setEventIds(eventIds);
+		dbMonitoring.getEventInfoList(eventList, eventOption);
+		EventInfo eventInfo = *eventList.begin();
```

Comment:
Normally, "eventList size is 0" is not occurred, but we have `hatohol-house-keeper` tool which breaks DB consistency between `action_logs` and `events`.
So, we should check data consistency here.